File: backend-springboot/src/main/java/com/vaultx/banking/entity/Transaction.java
package com.vaultx.banking.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "transactions")
public class Transaction {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @Enumerated(EnumType.STRING)
    @Column(name = "transaction_type")
    private TransactionType transactionType;

    @NotNull
    @Column(precision = 15, scale = 2)
    private BigDecimal amount;

    private String description;

    @Column(name = "recipient_account")
    private String recipientAccount;

    @Column(name = "balance_after", precision = 15, scale = 2)
    private BigDecimal balanceAfter;

    @Enumerated(EnumType.STRING)
    private TransactionStatus status;

    @Column(name = "reference_number", unique = true)
    private String referenceNumber;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    public enum TransactionType {
        CREDIT, DEBIT, TRANSFER
    }

    public enum TransactionStatus {
        PENDING, COMPLETED, FAILED, CANCELLED
    }

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (status == null) {
            status = TransactionStatus.PENDING;
        }
        if (referenceNumber == null) {
            referenceNumber = "TXN" + System.currentTimeMillis();
        }
    }

    // Constructors
    public Transaction() {}

    public Transaction(Customer customer, TransactionType transactionType, BigDecimal amount, String description) {
        this.customer = customer;
        this.transactionType = transactionType;
        this.amount = amount;
        this.description = description;
    }

    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public Customer getCustomer() { return customer; }
    public void setCustomer(Customer customer) { this.customer = customer; }

    public TransactionType getTransactionType() { return transactionType; }
    public void setTransactionType(TransactionType transactionType) { this.transactionType = transactionType; }

    public BigDecimal getAmount() { return amount; }
    public void setAmount(BigDecimal amount) { this.amount = amount; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public String getRecipientAccount() { return recipientAccount; }
    public void setRecipientAccount(String recipientAccount) { this.recipientAccount = recipientAccount; }

    public BigDecimal getBalanceAfter() { return balanceAfter; }
    public void setBalanceAfter(BigDecimal balanceAfter) { this.balanceAfter = balanceAfter; }

    public TransactionStatus getStatus() { return status; }
    public void setStatus(TransactionStatus status) { this.status = status; }

    public String getReferenceNumber() { return referenceNumber; }
    public void setReferenceNumber(String referenceNumber) { this.referenceNumber = referenceNumber; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
}

File: backend-dotnet/VaultX.Banking.API/Models/Transaction.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace VaultX.Banking.API.Models;

public class Transaction
{
    [Key]
    public int Id { get; set; }

    [Required]
    public int CustomerId { get; set; }

    [Required]
    public TransactionType TransactionType { get; set; }

    [Required]
    [Column(TypeName = "decimal(15,2)")]
    public decimal Amount { get; set; }

    [StringLength(255)]
    public string Description { get; set; } = string.Empty;

    [StringLength(50)]
    public string? RecipientAccount { get; set; }

    [Column(TypeName = "decimal(15,2)")]
    public decimal BalanceAfter { get; set; }

    public TransactionStatus Status { get; set; } = TransactionStatus.Pending;

    [StringLength(50)]
    public string ReferenceNumber { get; set; } = string.Empty;

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    // Navigation property
    [ForeignKey("CustomerId")]
    public virtual Customer Customer { get; set; } = null!;
}

public enum TransactionType
{
    Credit,
    Debit,
    Transfer
}

public enum TransactionStatus
{
    Pending,
    Completed,
    Failed,
    Cancelled
}
